pkg/providers/docker/controlplane_test.go

```diff
@@ -0,0 +1,527 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	dockerv1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/eks-anywhere/internal/test"
+	"github.com/aws/eks-anywhere/pkg/clients/kubernetes"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/providers/docker"
+)
+
+const (
+	testClusterConfigMainFilename = "testdata/cluster_main.yaml"
+)
+
+type baseControlPlane = clusterapi.ControlPlane[*dockerv1.DockerCluster, *dockerv1.DockerMachineTemplate]
+
+func TestControlPlaneObjects(t *testing.T) {
+	tests := []struct {
+		name         string
+		controlPlane *docker.ControlPlane
+		want         []kubernetes.Object
+	}{
+		{
+			name: "stacked etcd",
+			controlPlane: &docker.ControlPlane{
+				BaseControlPlane: baseControlPlane{
+					Cluster:                     capiCluster(),
+					ProviderCluster:             dockerCluster(),
+					KubeadmControlPlane:         kubeadmControlPlane(),
+					ControlPlaneMachineTemplate: dockerMachineTemplate("cp-mt"),
+				},
+			},
+			want: []kubernetes.Object{
+				capiCluster(),
+				dockerCluster(),
+				kubeadmControlPlane(),
+				dockerMachineTemplate("cp-mt"),
+			},
+		},
+		{
+			name: "unstacked etcd",
+			controlPlane: &docker.ControlPlane{
+				BaseControlPlane: baseControlPlane{
+					Cluster:                     capiCluster(),
+					ProviderCluster:             dockerCluster(),
+					KubeadmControlPlane:         kubeadmControlPlane(),
+					ControlPlaneMachineTemplate: dockerMachineTemplate("cp-mt"),
+					EtcdCluster:                 etcdCluster(),
+					EtcdMachineTemplate:         dockerMachineTemplate("etcd-mt"),
+				},
+			},
+			want: []kubernetes.Object{
+				capiCluster(),
+				dockerCluster(),
+				kubeadmControlPlane(),
+				dockerMachineTemplate("cp-mt"),
+				etcdCluster(),
+				dockerMachineTemplate("etcd-mt"),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			g.Expect(tt.controlPlane.Objects()).To(ConsistOf(tt.want))
+		})
+	}
+}
+
+func TestControlPlaneSpecNewCluster(t *testing.T) {
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	client := test.NewFakeKubeClient()
+	spec := test.NewFullClusterSpec(t, testClusterConfigMainFilename)
+
+	cp, err := docker.ControlPlaneSpec(ctx, logger, client, spec)
+
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(cp).NotTo(BeNil())
+	g.Expect(cp.Cluster).To(Equal(capiCluster()))
+	g.Expect(cp.KubeadmControlPlane).To(Equal(kubeadmControlPlane()))
+	g.Expect(cp.EtcdCluster).To(Equal(etcdCluster()))
+	g.Expect(cp.ProviderCluster).To(Equal(dockerCluster()))
+	g.Expect(cp.ControlPlaneMachineTemplate.Name).To(Equal("test-control-plane-1"))
+	g.Expect(cp.EtcdMachineTemplate.Name).To(Equal("test-etcd-1"))
```

Comment:
Can we assert the content of the machine templates?

Reply:
Yes., I've made the changes